### PR TITLE
Pagination

### PR DESCRIPTION
**Pull request by LindsayYoung**
[pagination](https://app.zenhub.com/workspaces/doj-crt-intake-scrum-board-5d03bf56c1c8a35d482eca0f/issues/18f/crt-portal/28)

## What does this change?
Adds pagination functionality. Adds some bare-bones styling. Adds roles and tags and labels for accessibility. 

## Screenshots (for front-end PR):
Screen Shot 2019-11-07 at 8 16 46 PM

## Checklist:
Just looking for the functional ACs for [pagination](https://app.zenhub.com/workspaces/doj-crt-intake-scrum-board-5d03bf56c1c8a35d482eca0f/issues/18f/crt-portal/28)

Style will follow.

### Author

+ [x] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.)
+ [x] Check for [accessibility](/docs/a11y_plan.md).
+ [x] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

### Reviewer
+ [ ] Ran CRT-view a11y check and it passed
+ [ ] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.)
+ [ ] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

## Notes for reviewer:
add `?per_page=1` or whatever number to the url, that makes it easier to test multiple pages.

There is a case where it shows ... when it doesn't need it  so: 1 ... 2 3 4 ... 100. I am not sure if it is worth cleaning up that edge case since it's not wrong. 

See [PR instructions doc](https://github.com/usdoj/crt-portal/blob/master/docs/pull_requests.md) for full pull request review instructions.

### CODE REVIEW

**Comment by LindsayYoung** [comment]
@alexsoble this is good to go

File: crt_portal/cts_forms/views.py [hunk 1]
```diff
@@ -33,7 +41,9 @@ def IndexView(request):
             "report_protected_classes": p_class_list
         })
 
-    return render_to_response('forms/index.html', {'data_dict': data})
+        print(page_format)
```
Comment (alexsoble) [inline crt_portal/cts_forms/views.py]: Noticing a stray `print` statement here.

File: crt_portal/cts_forms/tests.py [hunk 2]
```diff
@@ -114,6 +115,44 @@ def test_violation_summary(self):
...
+        # we are making a test record for each protected class
+        num_records = len(PROTECTED_CLASS_CHOICES)
+        self.assertTrue(f'{num_records} of {num_records} records' in self.content)
+
+    def test_paging(self):
+        url_base = reverse('crt_forms:crt-forms-index')
+        url = f'{url_base}?page=6&per_page=1'
+        response = self.client.get(url)
+        content = str(response.content)
+        print(content)
```
Comment (alexsoble) [inline crt_portal/cts_forms/tests.py]: Do we want this `print` statement here?

File: crt_portal/cts_forms/tests.py [hunk 3]
```diff
@@ -114,6 +115,44 @@ def test_violation_summary(self):
...
+        # check first page, current page, and the pages before and after
+        self.assertTrue("Go to page 1" in content)
```
Comment (alexsoble) [inline crt_portal/cts_forms/tests.py]: Looks like this only tests the state on page load and doesn't exercise the functionality of clicking through the links, that would be a nice-to-have (though not a blocker).

File: crt_portal/cts_forms/templates/forms/snippets/pagination.html [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<nav class="usa-pagination" role="navigation" aria-label="Pagination">
```
Comment (alexsoble) [inline crt_portal/cts_forms/templates/forms/snippets/pagination.html]: This is great, seeing all these great a11y reccommendations from a11ymatters in action! 🙌 

https://www.a11ymatters.com/pattern/pagination/

File: crt_portal/cts_forms/templates/forms/snippets/pagination.html [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  {% if page_format.show_first %}
+    <li><a href="?page=1&per_page={{ page_format.per_page }}" role="button" aria-label="Go to page 1">1</a></li>
+  {% endif %}
+  {% if page_format.first_ellipsis %}
+    <li><span class="ellipsis" aria-hidden='true'>...</span></li>
+  {% endif %}
+
+  {% for linkpage in page_format.page_numbers %}
+    {% ifequal linkpage page_format.page %}
+       <li class="is-active" aria-label="Current page, page {{ page_format.page }}"><b>{{ page_format.page }}</b></li>
```
Comment (alexsoble) [inline crt_portal/cts_forms/templates/forms/snippets/pagination.html]: I think this should have `aria-current="true"` set: 

https://www.a11ymatters.com/pattern/pagination/#3--indicate-which-element-is-currently-active

If you want to update this now, excellent, or feel free to leave it to me when I pick up styling if that's easier. 🚢

File: crt_portal/cts_forms/templates/forms/snippets/pagination.html [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    {% else %}
+       <li><a href="?page={{ linkpage }}&per_page={{ page_format.per_page }}" role="button" aria-label="Go to page {{ linkpage }}">{{ linkpage }}</a></li>
```
Comment (alexsoble) [inline crt_portal/cts_forms/templates/forms/snippets/pagination.html]: I don't know if this is a real VO issue, but when I navigate through the options using VO+➡️, it sounds confusing:

"Go to page 7 10 of 11"

vo-issue-maybe

Adding puctuation might help? To be clear, this may be normal VO behavior, it just sounds a bit strange/unclear to my ears. Feel free to update this or not, but wanted to mention

File: crt_portal/cts_forms/page_through.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    adjacent_pages = 1
+
+    show_first = False
+    first_ellipsis = False
+    if page > adjacent_pages + 1:
+        show_first = True
+        first_ellipsis = True
+        if page <= adjacent_pages + 2:
+            first_ellipsis = False
```
Comment (alexsoble) [inline crt_portal/cts_forms/page_through.py]: Could we simplify this logic down to two lines? Something like:

```
show_first = page > adjacent_pages + 1
first_ellipsis = show_first and not (page <= adjacent_pages + 2)
```

File: crt_portal/cts_forms/tests.py (hunk 3 above)
Comment (LindsayYoung) [reply crt_portal/cts_forms/tests.py]: Since the clicks are just links, we can test a link to check they are generating correctly. I think that is good enough for now

File: crt_portal/cts_forms/templates/forms/snippets/pagination.html (hunk 6 above)
Comment (LindsayYoung) [reply crt_portal/cts_forms/templates/forms/snippets/pagination.html]: adding punctuation

File: crt_portal/cts_forms/page_through.py (hunk 7 above)
Comment (LindsayYoung) [reply crt_portal/cts_forms/page_through.py]: Thanks, great suggestion!!!

File: crt_portal/cts_forms/page_through.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    except EmptyPage:
+        records = paginator.page(paginator.num_pages)
+
+    adjacent_pages = 1
+
+    show_first = page > adjacent_pages + 1
+    first_ellipsis = show_first and not (page <= adjacent_pages + 2)
+
+    show_last = page < paginator.num_pages - adjacent_pages
+    last_ellipsis = show_last and not (paginator.num_pages - adjacent_pages - 1)
```
Comment (alexsoble) [inline crt_portal/cts_forms/page_through.py]: @LindsayYoung Should this line be as follows?:

```
last_ellipsis = show_last and not (page == paginator.num_pages - adjacent_pages - 1)
```

Comment (LindsayYoung) [reply crt_portal/cts_forms/page_through.py]: Yes, thank you!! I can update in the next 20 minutes or so

**Comment by LindsayYoung** [comment]
We can remove the comment, I just added it because it seemed a bit dense. 

> @LindsayYoung I'm noticing that the ellipsis seems off after the refactor, and I think I see why — will highlight where I think there's a bug using an inline code comment.
> 
> ellipsis
